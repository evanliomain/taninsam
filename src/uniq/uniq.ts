/**
 * @module array=>array
 */
import { chain } from '../chain';
import { hash } from '../hash';
import { isUndefined } from '../is-undefined';
import { reduce } from '../reduce';
/**
 * Returns a new list containing only one copy of each element
 * in the original list, based upon the value returned by applying the supplied
 * function to each list element.
 * Prefers the first item if the supplied function produces
 * the same value on two items.
 * === is used for comparison.
 * @param comparison The comparison method, default set to identity.
 * @return the function to apply on the array to uniq its elements
 * @example
 * ```
 * uniq<number>()([1, 2, 3, 2, 3, 5, 1]) // [1, 2, 3, 5]
 * uniq<{ a: number; b: number; c: string; }>(({ a, b, c }) => ({ a, b }))([
 *   { a: 1, b: 1, c: 'a' },
 *   { a: 1, b: 2, c: 'a' },
 *   { a: 1, b: 1, c: 'b' }
 * ])) // [{ a: 1, b: 1, c: 'a' }, { a: 1, b: 2, c: 'a' }]
 * ```
 * @example Using the chain
 * ```
 * chain([
 *   { a: 1, b: 1, c: 'a' },
 *   { a: 1, b: 2, c: 'a' },
 *   { a: 1, b: 1, c: 'b' }
 * ])
 * .chain(uniq<{ a: number; b: number; c: string; }>(
 *  ({ a, b, c }) => ({ a, b }))
 * )
 * .value() // [{ a: 1, b: 1, c: 'a' }, { a: 1, b: 2, c: 'a' }]
 * ```
 */
export function uniq<T>(
  comparison: (x: T) => any = x => x,
): (element: T[]) => T[] {
  return (array: T[]) =>
    chain(array)
      .chain(
        reduce(
          (
            acc: { result: T[]; distinct: { [key: string]: boolean } },
            element: T,
          ) => {
            const value = comparison(element);
            const hashKey = hash()(value);
            if (!isUndefined(acc.distinct[hashKey])) {
              return acc;
            }

            return {
              result: [...acc.result, element],
              distinct: { ...acc.distinct, [hashKey]: true },
            };
          },
          { result: [], distinct: {} },
        ),
      )
      .chain(({ result }) => result)
      .value();
}
