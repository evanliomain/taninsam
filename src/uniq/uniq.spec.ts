import { describe, expect, test } from 'vitest';
import { uniq } from './uniq';

type Sample = {
  a: number;
  b: number;
  c: string;
};

describe('uniq function', () => {
  const input: Sample[] = [
    { a: 1, b: 1, c: 'a' },
    { a: 1, b: 2, c: 'a' },
    { a: 1, b: 1, c: 'b' }
  ];

  test('is a pure function', () => {
    uniq()(input);
    expect(input).toMatchSnapshot();
  });

  test('[1, 2, 3, 2, 3, 5, 1] |> uniq === [1, 2, 3, 5]', () => {
    expect(uniq<number>()([1, 2, 3, 2, 3, 5, 1])).toMatchSnapshot();
  });

  test('input |> uniq ({ a, b, c }) => a === [{ a: 1, b: 1, c: "a" }]', () => {
    expect(uniq<Sample>(({ a }) => a)(input)).toMatchSnapshot();
  });

  test('input |> uniq ({ a, b, c }) => ({ a, b }) === [{ a: 1, b: 1, c: "a" }, { a: 1, b: 2, c: "a" }]', () => {
    expect(uniq<Sample>(({ a, b }) => ({ a, b }))(input)).toMatchSnapshot();
  });

  test('input |> uniq ({ a, b, c }) => ({ a, c }) === [{ a: 1, b: 1, c: "a" }, { a: 1, b: 1, c: "b" }]', () => {
    expect(uniq<Sample>(({ a, c }) => ({ a, c }))(input)).toMatchSnapshot();
  });

  test('deepObject |> uniq', () => {
    expect(
      uniq()([
        { a: { b: 1 }, b: { a: 1 } },
        { a: { b: 1 }, b: { a: 2 } },
        { a: { b: 2 }, b: { a: 1 } },
        { a: { b: 1 }, b: { a: 1 } }
      ])
    ).toMatchSnapshot();
  });
});
