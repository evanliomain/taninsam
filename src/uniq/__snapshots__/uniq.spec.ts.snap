// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uniq function > [1, 2, 3, 2, 3, 5, 1] |> uniq === [1, 2, 3, 5] 1`] = `
[
  1,
  2,
  3,
  5,
]
`;

exports[`uniq function > deepObject |> uniq 1`] = `
[
  {
    "a": {
      "b": 1,
    },
    "b": {
      "a": 1,
    },
  },
  {
    "a": {
      "b": 1,
    },
    "b": {
      "a": 2,
    },
  },
  {
    "a": {
      "b": 2,
    },
    "b": {
      "a": 1,
    },
  },
]
`;

exports[`uniq function > input |> uniq ({ a, b, c }) => ({ a, b }) === [{ a: 1, b: 1, c: "a" }, { a: 1, b: 2, c: "a" }] 1`] = `
[
  {
    "a": 1,
    "b": 1,
    "c": "a",
  },
  {
    "a": 1,
    "b": 2,
    "c": "a",
  },
]
`;

exports[`uniq function > input |> uniq ({ a, b, c }) => ({ a, c }) === [{ a: 1, b: 1, c: "a" }, { a: 1, b: 1, c: "b" }] 1`] = `
[
  {
    "a": 1,
    "b": 1,
    "c": "a",
  },
  {
    "a": 1,
    "b": 1,
    "c": "b",
  },
]
`;

exports[`uniq function > input |> uniq ({ a, b, c }) => a === [{ a: 1, b: 1, c: "a" }] 1`] = `
[
  {
    "a": 1,
    "b": 1,
    "c": "a",
  },
]
`;

exports[`uniq function > is a pure function 1`] = `
[
  {
    "a": 1,
    "b": 1,
    "c": "a",
  },
  {
    "a": 1,
    "b": 2,
    "c": "a",
  },
  {
    "a": 1,
    "b": 1,
    "c": "b",
  },
]
`;
